### DISCUSSION TITLE
[Language Modes] Add Further Coding Languages to Default Selection

### DISCUSSION THREAD

**Discussion by kevinmatthes** [zee-editor/zee#57] [Ideas]
Even though users are allowed to add further coding language modes themselves in their configuration files when needed, I would like to suggest to add some further language modes to the default configuration file anyway since there are many coding languages tree-sitters exist for.

This thread is intended to discuss and prioritise the adding of new tree-sitters to the default configuration file.

**Comment by kevinmatthes** [comment]
For which coding language(s) would you like to request support for by adding their tree-sitters to the default configuration file?  Add a checklist to this comment as a reply.

**Reply by kevinmatthes** [reply]
* [x] Haskell
* [ ] Fortran
* [ ] LaTeX
* [ ] Lex / Flex
* [ ] YACC / Bison

**Reply by kevinmatthes** [reply]
* [ ] Makefile
* [ ] CMake
* [ ] justfile
* [ ] MATLAB
* [ ] GNU Octave

**Reply by iainh** [reply]
In the default configuration: 
- [ ] yaml
- [ ] git-commit

In a secondary configuration that is either on the website or in the source tree for the user: 
- [ ] Java
- [ ] Zig

**Reply by kevinmatthes** [reply]
* [ ] CFF
* [ ] Godot Script

**Comment by iainh** [comment]
https://github.com/helix-editor/helix/blob/master/languages.toml has a fairly comprehensive list of what is generally available and I wrote a quick application to ingest languages.toml and output a config.ron when I was testing mode loading earlier. Some language implementations seem have long load times (looking at you swift) so I'd like to propose being conservative in what languages get included by default but make it easy for users to discover additional languages and customize the list. Perhaps the website could contain a curated list or the default configuration could include everything but with most commented out to begin with.

**Reply by kevinmatthes** [reply]
The idea of the application you mentioned leads to another option:  directory based mode integration.

At the moment, the configuration directory of Zee has subdirectories for the tree-sitters and their update mechanism.  When we give this configuration directory another subdirectory for the modes, users could store their language specific RON configurations there.

The main configuration file would then just need a path to this directory in order to load the language modes.  Sample configurations could be provided by a repository on its own within this namespace in order enable users to clone it directly into their configuration directories for simple update fetching.  If users would like to deactivate a certain mode they just need to remove the respective RON file.

**Reply by kevinmatthes** [reply]
@iainh, is your application to extract the modes from the configuration file available somewhere?

**Reply by iainh** [reply]
It isn't currently know. I was mainly a quick hack to get a as many modes as I could for some testing rather than something I think anyone should use.

**Comment by kevinmatthes** [comment]
I just added Haskell support with #62.

**Reply by kevinmatthes** [reply]
The Haskell mode was merged with #75.

**Comment by kevinmatthes** [comment]
Should we provide a tool to create new modes more easily?

**Reply by kevinmatthes** [reply]
As @iainh mentioned, the process of defining new language modes can be automated.  I would like to suggest to create a tool for the command line to create the entry for the configuration more easily since, when doing so by hand, one often notices complaints that some fields are either missing or are not formatted appropriately.

For instance, the tool could take the URL to a GitHub repository with a Tree-Sitter and then fetch the latest revision's hexadecimal identifier to add it in the corresponding field.  Other relevant data could be set by command line options.  In the end, a RON file with the determined data could be written in the working directory.

Would such a tool be useful?

**Comment by kevinmatthes** [comment]
The VS Code-like IDEs allow C-style comments for JSON (JavaScript Object Notation).  One can choose whether to use default JSON or the version with comments with the latter version being the default setting.

Should this be provided in the setup of Zee, as well?

**Comment by kevinmatthes** [comment]
I suggested the integration for the **Just Build System** in #87.  This Pull Request also proposes an emergency override mechanism for the default path to the important `*.scm` files which will help to add further languages.
